packages/idempotency/tests/unit/persistence/DynamoDbPersistenceLayer.test.ts

```diff
@@ -29,31 +34,77 @@ describe('Class: DynamoDbPersistenceLayer', () => {
     public _updateRecord(record: IdempotencyRecord): Promise<void> {
       return super._updateRecord(record);
     }
+
   }
 
   beforeEach(() => {
     jest.clearAllMocks();
   });
 
+  describe('Method: constructor', () => {
+
+    test('when instantiated with minimum options it creates an instance with default values', () => {
+
+      // Prepare
+      const persistenceLayer = new TestDynamoDBPersistenceLayer({ tableName: dummyTableName });
+
+      // Act / Assess
+      expect(persistenceLayer).toEqual(expect.objectContaining({
+        tableName: dummyTableName,
+        keyAttr: 'id',
+        statusAttr: 'status',
+        expiryAttr: 'expiration',
+        inProgressExpiryAttr: 'in_progress_expiry_attr',
+        dataAttr: 'data'
+      }));
+
+    });
+
+    test('when instantiated with specific options it creates an instance with correct values', () => {
+
+      // Prepare
+      const testDynamoDBPersistenceLayerOptions: DynamoPersistenceConstructorOptions = {
+        tableName: dummyTableName,
+        keyAttr: dummyKey,
+        statusAttr: 'someStatusAttr',
+        expiryAttr: 'someExpiryAttr',
+        inProgressExpiryAttr: 'someInProgressExpiryAttr',
+        dataAttr: 'someDataAttr'
+      };
+      const persistenceLayer = new TestDynamoDBPersistenceLayer(testDynamoDBPersistenceLayerOptions);
+
+      // Act / Assess
+      expect(persistenceLayer).toEqual(expect.objectContaining({
+        tableName: dummyTableName,
+        keyAttr: dummyKey,
+        statusAttr: testDynamoDBPersistenceLayerOptions.statusAttr,
+        expiryAttr: testDynamoDBPersistenceLayerOptions.expiryAttr,
+        inProgressExpiryAttr: testDynamoDBPersistenceLayerOptions.inProgressExpiryAttr,
+        dataAttr: testDynamoDBPersistenceLayerOptions.dataAttr
+      }));
+
+    });
+
+  });
+
```

Comment:
All member variables in this class are private. Should we break the abstraction here?

<img width="541" alt="image" src="https://user-images.githubusercontent.com/61999/201927824-c7362bdc-aa20-46d6-bb15-562bd5271732.png">

The private value are indirectly tested in the `_putRecord()` and other protected fields. Those test cases would detect an issue if the private values are incorrect.

Reply:
I see your point.

The reason why I introduced these tests is primarily to reach 100% coverage.

If we remove this test, coverage drops:
![image](https://user-images.githubusercontent.com/7353869/201974561-2c65211a-77b0-4265-9c1a-7f153fe4420b.png)

and the reason, I think, is because we have conditional statements in the `constructor`:
```ts
this.tableName = constructorOptions.tableName;
this.keyAttr = constructorOptions.keyAttr ?? 'id';
this.statusAttr = constructorOptions.statusAttr ?? 'status';
this.expiryAttr = constructorOptions.expiryAttr ?? 'expiration';
this.inProgressExpiryAttr = constructorOptions.inProgressExpiryAttr ?? 'in_progress_expiry_attr';
this.dataAttr = constructorOptions.dataAttr ?? 'data';
```

Since in all other test cases, including `_putRecord` we are always using the defaults, we never testing the left-side of the `??` (except for `tableName`).

Now, I could have an additional test to one of the other methods (i.e. `_putRecord`) and pass some values when initializing the class, however I saw that test as really testing the constructor, and so I added its own case.

If there's another way I'm happy to change it though, I have to admit I went this way also because this is how I've done it in Tracer.